Bring FPGA-Test-Shield tests into Mbed-os master.

mprse opened the pull request:
### Description

Bring FPGA-Test-Shield tests to Mbed-os. 

When this is merged these tests can be removed from `ci-tes-shield-fpga` repository.

By default, all these tests will be skipped on CI. To enable these tests `FPGA_CI_TEST_SHIELD` component must be added in targets.json configuration file.

Related PRs: PR https://github.com/ARMmbed/mbed-os/pull/10816

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [X] Test update
    [ ] Breaking change

### Reviewers

@maciejbocianski 
@fkjagodzinski 
@c1728p9 
@jamesbeyond

mprse committed:
Bring FPGA-Test-Shield tests into Mbed-os master.

jamesbeyond reviewed (commented):
These tests are HAL tests, shall we add the header files for the test along with doxygen comments. (same as rest of the hal tests)
@mprse @maciejbocianski @fkjagodzinski @c1728p9

c1728p9 commented on TESTS/mbed_hal_fpga_ci_test_shield/i2c/main.cpp:
```diff
@@ -0,0 +1,328 @@
+#error [NOT_SUPPORTED] I2C not supported for this target
+#elif !COMPONENT_FPGA_CI_TEST_SHIELD
+#error [NOT_SUPPORTED] FPGA CI Test Shield is needed to run this test
+#elif !defined(FULL_TEST_SHIELD)
```
`FULL_TEST_SHIELD` should probably be removed since this only applies to the Basys3.

mprse replied:
Fixed

mprse committed:
FPGA I2C test update.

Divide one test case to four test cases to increase readability.
Adapted to the last version of the FPGA CI Test Shield API.

mprse committed:
FPGA Analogin test: Remove usage of FULL_TEST_SHIELD symbol, increase tolerance

Keep "AnalogIn - full test" disabled due to hardware issue in the first rev of FPGA Test Shield.

0xc0170 commented on TESTS/mbed_hal_fpga_ci_test_shield/analogout/main.cpp:
```diff
@@ -0,0 +1,126 @@
+    // printf("Sum: %llu\r\n", sum);
+    // printf("Num power samples: %d\r\n", samples);
+    // printf("Num power cycles: %llu\r\n", cycles);
+    // printf("ANIN0 voltage: %.6fV\r\n", tester.get_anin_voltage(0));
```
Can we remove these lines not used?

mprse replied:
I think that this code has been added for debugging. Maybe I can wrap this code around `#if DEBUG` directive?

mprse replied:
Fixed

0xc0170 reviewed (commented):
Before I approve, just one comment about dead code below.

Where is the documentation for fpga testing (reference for docs PR?), plus what these are testing - I am seeing just test files here and defined test cases but where is the test plan (or something similar) or is all suficient what we have in Cases() in the files?

mprse committed:
FPGA Analogout test: mark dead code as debug code, increase tolerance

jamesbeyond reviewed (approved):
After a offline conversation,  I'd agree that because we don't have defined behavior for current these HAL APIs, the test header files can come to the repo as separate PR

0xc0170 commented on TESTS/mbed_hal_fpga_ci_test_shield/uart/main.cpp:
```diff
@@ -0,0 +1,353 @@
+    Case("115200, 8N1, FC on", one_peripheral<UARTPort, DefaultFormFactor, test_common<115200, 8, ParityNone, 1> >),
+    Case("115200, 8N1, FC off", one_peripheral<UARTNoFCPort, DefaultFormFactor, test_common_no_fc<115200, 8, ParityNone, 1> >), // DISCO_L475VG_IOT01A
+    // data bits
+    /* FIXME not supported: K64F
```
do we need to keep this dead code here? Can it be tracked separately (removed from here, stashed somewhere else with the tracking issue for this)

mprse replied:
@fkjagodzinski Can you check this?

fkjagodzinski replied:
Ok, so I checked this part -- data bits on K64F.

1. Our HAL allows setting `data_bits` and does not restrict this value in any way:
https://github.com/ARMmbed/mbed-os/blob/b55344c50b5602ee3e6c0cce748178bede3153ea/hal/serial_api.h#L137-L144
1. The HAL implementation for K64F does not really use `data_bits` param:
https://github.com/ARMmbed/mbed-os/blob/b55344c50b5602ee3e6c0cce748178bede3153ea/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/serial_api.c#L102
1. ... and this is because the `fsl_uart.c` does not have any `data_bits`-related config item:
https://github.com/ARMmbed/mbed-os/blob/b55344c50b5602ee3e6c0cce748178bede3153ea/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/drivers/fsl_uart.h#L167-L181
1. However, the `K64 Sub-Family Reference Manual` says there are two settings:
```
Programmable 8-bit or 9-bit data format
```

Initially I thought we should uncomment this block of code and expect NXP to provide a fix. Now, knowing NXP would have to update their HAL, I don't think it would be reasonable to do so. Also, I'm not so sure if Mbed OS should require support of 9-bit UART data at all.

@jamesbeyond, what do you think about this? Should I remove the `data_bits` part of this test?
BTW, there are two more similar `FIXME` comments regarding ST targets. Remove these as well? (don't know why they fail yet).

0xc0170 replied:
I vote for yes. dead code should be removed (tracking issue to have this, with commit including the changes needed). It will be easy to bring this back once it is fixed

mprse replied:
Fixed (d2d3d5f35db56ccf5aaddf7e7ceb59f950d3908a)

0xc0170 commented on TESTS/mbed_hal_fpga_ci_test_shield/analogin/main.cpp:
```diff
@@ -0,0 +1,167 @@
+    Case("AnalogIn - init test", all_ports<AnaloginPort, DefaultFormFactor, analogin_init>),
+    // This test case is disabled for now due to hardware issue in the first rev of FPGA Test Shield
+#if 0
+    Case("AnalogIn - full test", all_ports<AnaloginPort, DefaultFormFactor, analogin_full_test>),
```
remove as well this one not used

mprse replied:
This one will be enabled as soon as we get the rev.2 of the FPGA-Test-Shield and I will be able to rework and execute this test case.

0xc0170 replied:
It should come with rv2 (we should not have if 0 in the code)

mprse replied:
Ok, I will remove this test case for now and add it later.

mprse replied:
fixed (879d8917c4e33b8d52d021727e39f557cd170b49)

0xc0170 commented on TESTS/mbed_hal_fpga_ci_test_shield/pwm/main.cpp:
```diff
@@ -0,0 +1,224 @@
+#include "test_utils.h"
+
+
+#define pwm_debug_printf(...)
```
Empty?

mprse replied:
This is macro for debugging. By default, it does nothing.

0xc0170 replied:
OK, I was expecting if debug , it does something

mprse committed:
FPGA Analogin test: Remove test case which can not be executed

The test case which checks full range cannot be executed at the moment due to a hardware bug in FPGA-Test-Shield.
This test case will be restored when the final version of FPGA-Test-Shield is ready.

mprse committed:
FPGA uart test: Remove unsupported test cases

Not all uart features are supported by specific platforms.
